### PR TITLE
Implement unhappy RP flows with linting and tests

### PR DESCRIPTION
**Pull request by giovanni-riela**
#### Motivation and Context

The goal of this PR is to significantly enhance the robustness of the Wallet Conformance Test suite by implementing comprehensive "unhappy flows" (negative tests) for the OID4VP presentation protocol. By verifying how the system handles invalid authorization requests, tampered redirect URIs, and malformed JARM payloads, we ensure that both the Wallet and Relying Party (RP) implementations strictly adhere to security requirements and protocol edge cases.

  In addition to the new test coverage, this PR introduces several architectural improvements and infrastructure updates:
   - New Conformance Tests: Added detailed validation for Authorization Requests (`RPR_001`, `RPR_002`, etc.) and Redirect URI responses (`RPR_020`, `RPR_041`, etc.).
   - External Trust Anchor Support: Implemented logic to allow the test suite to register with and fetch subordinate statements from an external Trust Anchor, expanding the tool's versatility for different federation environments.
   - Refactoring & Standardization: Renamed orchestrator result fields for better clarity (e.g., `*Result` to `*Response`).
   
The rp unhappy flows are currently disabled for the E2E tests github actions due to inability to run parallel to the happy flow as is, we are investigating solutions.

JIRA Ticket: WLEO-1159

### CODE REVIEW

File: .github/workflows/conformance-tests-e2e.yml [hunk 1]
```diff
@@ -85,5 +85,5 @@ jobs:
       # ── Presentation tests ───────────────────────────────────────────────────
       - name: Run presentation tests
         shell: bash
-        run: wct test:presentation
+        run: pnpm test:presentation --tests happy
```
Comment (flaviodelgrosso) [inline .github/workflows/conformance-tests-e2e.yml]: Does calling wct not working passing --tests flag in CI?

Comment (giovanni-riela) [reply .github/workflows/conformance-tests-e2e.yml]: Need to merge main into this branch. Curently the command line options are parsed only when calling `wct ...`.

Comment (giovanni-riela) [reply .github/workflows/conformance-tests-e2e.yml]: Correction: pnpm allows specifying the test to run, the `--tests' threw me off, we just pass the name as argument, not as an option

Comment (flaviodelgrosso) [reply .github/workflows/conformance-tests-e2e.yml]: I think we should implement a fix to keep wct command in CI, so we can check for regressions by executing it directly for every OS matrix target

File: tests/conformance/presentation/redirect-uri.presentation.spec.ts [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+  // -----------------------------------------------------------------------
+
+  test("RPR_109: Authorization Error Response encoding | RP parses authorization errors over application/x-www-form-urlencoded", async () => {
+    const log = baseLog.withTag("RPR_109");
+    const DESCRIPTION =
+      "RP parses authorization generic errors over x-www-form-urlencoded";
+    log.start("Conformance test: Authorization Error Response encoding");
+
+    let testSuccess = false;
+    try {
```
Comment (manuraf) [inline tests/conformance/presentation/redirect-uri.presentation.spec.ts:345]: The assertions here (and in RPR_109) are vacuous — `response.text()` always returns a `string`, so `expect(body).toBeDefined()` will pass even when the RP crashes with a 500 or returns a completely wrong status. A correct implementation and a broken one are indistinguishable.

Per OpenID4VP §8.2 and the IT-Wallet Remote Flow spec, when the wallet sends a valid OAuth 2.0 error response (e.g. `access_denied`), the RP **must** reply with HTTP **200** and `Content-Type: application/json`. You could replace the current checks with spec-mandated assertions:

```typescript
// RPR_108 — wallet sends explicit error response
expect(response.status, "RP must respond 200 to wallet error responses per OID4VP §8.2").toBe(200);
const contentType = response.headers.get("content-type") ?? "";
expect(
  contentType.includes("application/json"),
  "RP must return application/json for wallet error responses"
).toBe(true);

// Optionally verify no stack trace leakage
const body = await response.text();
expect(body.includes("at Object.<anonymous>"), "RP must not leak stack traces").toBe(false);
```

Same fix applies to RPR_109.

File: tests/conformance/presentation/authorization-request.presentation.spec.ts [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+  beforeAll(async () => {
+    const ctx = await orchestrator.runThroughAuthorize();
+
+    verifierMetadata = ctx.verifierMetadata;
+    walletAttestationResponse = ctx.walletAttestationResponse;
+    credentials = ctx.credentials;
+
+    // Run a fresh authorization step to capture a valid JARM + responseUri
+    const authResult = ctx.authorizationRequestResponse;
+    validResponseUri = authResult.response!.responseUri;
```
Comment (manuraf) [inline tests/conformance/presentation/authorization-request.presentation.spec.ts]: The non-null assertion (`!`) on `authResult.response` here — and the same pattern in `redirect-uri.presentation.spec.ts` — is technically safe as long as `assertStepSuccess` always throws on failure. However, the type system doesn't enforce that guarantee, and a degenerate RP returning `success: true` with no `response` body would silently propagate `undefined` into all shared variables. Every downstream test would then fail with `"Failed to fetch: undefined"` rather than a clear setup-failure message.

You could replace the `!` with an explicit guard that both narrows the type correctly and surfaces a descriptive error when setup goes wrong:

```typescript
// authorization-request.presentation.spec.ts — beforeAll
const ctx = await orchestrator.runThroughAuthorize();
verifierMetadata = ctx.verifierMetadata;
walletAttestationResponse = ctx.walletAttestationResponse;
credentials = ctx.credentials;

const authResponse = ctx.authorizationRequestResponse.response;
if (!authResponse) {
  throw new Error(
    "Setup failed: authorizationRequestResponse.response is undefined — RP did not return a valid authorization response",
  );
}
validResponseUri = authResponse.responseUri;
validJarmJwe = authResponse.authorizationResponse.jarm.responseJwe;
```

The same fix applies to `redirect-uri.presentation.spec.ts:33-39`. This way TypeScript narrows `authResponse` without needing `!`, and any setup failure produces a meaningful diagnostic rather than a `TypeError` deep in the test run.

File: tests/conformance/presentation/authorization-request.presentation.spec.ts [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+      verifierMetadata,
+      walletAttestation: attestationOverride ?? walletAttestationResponse,
+    });
+  }
+
+  // -----------------------------------------------------------------------
+  // Helper: POST to response_uri with custom options
+  // -----------------------------------------------------------------------
+
+  async function postToResponseUri(
```
Comment (manuraf) [inline tests/conformance/presentation/authorization-request.presentation.spec.ts]: The `postToResponseUri` function is byte-for-byte identical here (lines 80–94) and in `redirect-uri.presentation.spec.ts` (lines 63–77). Any future change — adding a common header, adjusting the timeout, or wiring in logging — would have to be made in two places and will eventually drift.

Since the project already has a `tests/helpers/` directory with established shared helpers (`par-validation-helpers.ts`, `use-test-summary.ts`), you could extract this into a new `tests/helpers/http-helpers.ts` and import it in both spec files via `#/helpers/http-helpers`. As a bonus, it would also let you clean up RPR_060, which currently bypasses the helper and calls `fetchWithConfig` directly.

```typescript
// tests/helpers/http-helpers.ts
import { fetchWithConfig, loadConfigWithHierarchy } from "@/logic";

export async function postToResponseUri(
  responseUri: string,
  body: string,
  options?: { contentType?: string; method?: string },
): Promise<Response> {
  const config = loadConfigWithHierarchy();
  return fetchWithConfig(config.network)(responseUri, {
    body,
    headers: {
      "Content-Type":
        options?.contentType ?? "application/x-www-form-urlencoded",
    },
    method: options?.method ?? "POST",
  });
}
```

File: tests/conformance/presentation/authorization-request.presentation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,613 @@
...
+      const response = await postToResponseUri(
+        responseUri,
+        formBody.toString(),
+      );
+
+      log.debug(`  Response status: ${response.status}`);
+      log.info("→ Validating RP rejected the nonce mismatch...");
+      expect(response.ok).toBe(false);
+
+      testSuccess = true;
```
Comment (manuraf) [inline tests/conformance/presentation/authorization-request.presentation.spec.ts:339]: The current assertion here only verifies that the response is not successful, but the IT-Wallet conformance matrix specifically requires HTTP 403 when nonce validation fails. A 400 or 500 would also pass this check, making the test weaker than what the spec prescribes.

You could tighten this to pin the exact status code:

```typescript
expect(response.status, "RP must return 403 for nonce mismatch per IT-Wallet spec").toBe(403);
```

While you are at it, there are two related gaps in `redirect-uri.presentation.spec.ts` worth addressing together:

**RPR_020** (lines 113–119): The test appends a path suffix on the same RP host. This will produce a 404 because the route does not exist — not because of any redirect_uri security check. The spec-relevant scenario is a redirect to a host that was never attested in the Trust Chain. If host-level testing is not feasible (DNS/network constraints), consider documenting the limitation inline:

```typescript
// NOTE: This test verifies path-level rejection only. Host-level redirect_uri
// security (unattested host) cannot be tested without RP-side DNS control.
// See IT-Wallet spec RPR_020 for full scenario.
```

**RPR_114** (`redirect-uri.presentation.spec.ts:352`): `body.error_description` is currently not asserted, even though RFC 6749 §5.2 requires both `error` and `error_description`. RPR_099 in the same file already checks both — it would be consistent to do the same here:

```typescript
expect(body.error, "Validation error response must contain 'error' field").toBeDefined();
expect(
  body.error_description,
  "Validation error response must contain 'error_description' field per RFC 6749 §5.2",
).toBeDefined();
```

File: tests/conformance/presentation/authorization-request.presentation.spec.ts [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+  // -----------------------------------------------------------------------
+  // RPR-76 — Unsupported HTTP methods
+  // -----------------------------------------------------------------------
+
+  test("RPR_076: Unsupported HTTP methods | RP rejects PUT, DELETE, and PATCH requests to response_uri", async () => {
+    const log = baseLog.withTag("RPR_076");
+    const DESCRIPTION = "RP correctly rejected unsupported HTTP methods";
+    log.start("Conformance test: Unsupported HTTP methods");
+
+    let testSuccess = false;
```
Comment (manuraf) [inline tests/conformance/presentation/authorization-request.presentation.spec.ts]: Testing three methods sequentially with `expect` inside a `for...of` loop means the first failing assertion short-circuits the loop — if PUT passes but DELETE would fail, the DELETE and PATCH branches are never exercised, and the test result only reflects the first failure encountered.

You could restructure this by firing all three requests up front, collecting the results, and then asserting on all of them:

```typescript
const results = await Promise.all(
  unsupportedMethods.map(async (method) => ({
    method,
    response: await postToResponseUri(freshUri_for_method, formBody.toString(), { method }),
  })),
);
for (const { method, response } of results) {
  expect(response.ok, `RP should reject ${method} on response_uri`).toBe(false);
}
```

Worth noting: once issue #001 (one-time-use URI) is resolved, each iteration will need its own fresh `responseUri` — the `freshUri_for_method` placeholder above represents that. Alternatively, `test.each` would also work here if you'd prefer independent sub-tests per method, each with its own clear pass/fail in the reporter.

File: src/types/presentation-orchestrator-context.ts [hunk 7]
```diff
@@ -1,11 +1,24 @@
+import { ItWalletCredentialVerifierMetadata } from "@pagopa/io-wallet-oid-federation";
+
 import type { AuthorizationRequestStepResponse } from "@/step/presentation";
 import type { FetchMetadataVpStepResponse } from "@/step/presentation";
 import type { RedirectUriStepResponse } from "@/step/presentation";
```
Comment (manuraf) [inline src/types/presentation-orchestrator-context.ts]: Three separate `import type` lines from the same module could be merged into a single grouped import — it's a bit more readable and keeps the import block tidy:

```typescript
import type {
  AuthorizationRequestStepResponse,
  FetchMetadataVpStepResponse,
  RedirectUriStepResponse,
} from "@/step/presentation";
```

File: tests/conformance/presentation/authorization-request.presentation.spec.ts [hunk 8]
```diff
@@ -0,0 +1,613 @@
...
+        (k) => k.use === "enc",
+      );
+      expect(encryptionKey, "RP must have an encryption key").toBeDefined();
+
+      const malformedPayload = JSON.stringify({
+        state: "invalid-state",
+        vp_token: { invalid_credential_id: 12345 },
+      });
+
+      const encryptJwe = getEncryptJweCallback(encryptionKey!);
```
Comment (manuraf) [inline tests/conformance/presentation/authorization-request.presentation.spec.ts]: The pattern `verifierMetadata.jwks.keys.find(k => k.use === "enc")` followed by `expect(encryptionKey).toBeDefined()` and then `encryptionKey!` appears in four tests here (RPR_025, RPR_026, RPR_039, RPR_066). Worth noting: Vitest's `toBeDefined()` is not a TypeScript type predicate — it doesn't narrow the type, so `encryptionKey` stays `JsonWebKey | undefined` and the `!` is still required. If the key is actually missing, you get a runtime `TypeError` instead of a meaningful assertion failure.

You could extract a small helper that throws with a descriptive message and returns a narrowed `JsonWebKey` — eliminating the `!` at all call sites:

```typescript
// tests/helpers/http-helpers.ts (or a new presentation-helpers.ts)
export function getVerifierEncryptionKey(
  verifierMetadata: ItWalletCredentialVerifierMetadata,
): JsonWebKey {
  const key = verifierMetadata.jwks.keys.find((k) => k.use === "enc");
  if (!key) {
    throw new Error(
      "RP metadata does not contain an encryption key (use=enc) — " +
        "cannot build JARM for this test. Check RP JWKS endpoint.",
    );
  }
  return key;
}
```

Then every call site becomes simply:

```typescript
const encryptionKey = getVerifierEncryptionKey(verifierMetadata);
// No ! needed — key is JsonWebKey (not undefined)
```

This also follows the AGENTS.md guidance on extracting helpers for repeated patterns.

Comment (giovanni-riela) [reply tests/conformance/presentation/authorization-request.presentation.spec.ts]: I noticed the key is always taken from the verifierMetadata in the same way, meaning the key retrieved is always the same. I opted for a the definition of a test variable populated along with the other variables before all tests. A missing encryption key fails all tests in the file as it is caused by malformed metadata retrieved from the verifier. Let me know what you think.

Comment (giovanni-riela) [reply tests/conformance/presentation/authorization-request.presentation.spec.ts]: addressed in 7d1eb7e

File: src/types/presentation-orchestrator-context.ts (hunk 7 above)
Comment (giovanni-riela) [reply src/types/presentation-orchestrator-context.ts]: addressed in 7d1eb7e

File: tests/conformance/presentation/authorization-request.presentation.spec.ts (hunk 6 above)
Comment (giovanni-riela) [reply tests/conformance/presentation/authorization-request.presentation.spec.ts]: addressed in 7d1eb7e

File: tests/conformance/presentation/authorization-request.presentation.spec.ts (hunk 5 above)
Comment (giovanni-riela) [reply tests/conformance/presentation/authorization-request.presentation.spec.ts:339]: addressed in 7d1eb7e

File: tests/conformance/presentation/authorization-request.presentation.spec.ts (hunk 4 above)
Comment (giovanni-riela) [reply tests/conformance/presentation/authorization-request.presentation.spec.ts]: addressed in 7d1eb7e

File: tests/conformance/presentation/authorization-request.presentation.spec.ts (hunk 3 above)
Comment (giovanni-riela) [reply tests/conformance/presentation/authorization-request.presentation.spec.ts]: addressed in 7d1eb7e

File: tests/conformance/presentation/redirect-uri.presentation.spec.ts (hunk 2 above)
Comment (giovanni-riela) [reply tests/conformance/presentation/redirect-uri.presentation.spec.ts:345]: addressed in 7d1eb7e

**Review by manuraf — APPROVED** [review]
